src/twisted/test/test_defer.py

```diff
@@ -1678,6 +1681,42 @@ def test_fromCoroutineRequiresCoroutine(self) -> None:
         for thing in thingsThatAreNotCoroutines:
             self.assertRaises(defer.NotACoroutineError, Deferred.fromCoroutine, thing)
 
+    @pyunit.skipIf(_PYPY, "GC works differently on PyPy.")
+    def test_canceller_circular_reference(self) -> None:
+        """
+        Test that a circular reference between a `Deferred` and its canceller
+        is broken when the deferred is resolved.
+        """
+
+        # Create a canceller and weak reference to track if its been freed.
+        canceller = _TestCircularCanceller()
+        weakCanceller = weakref.ref(canceller)
+
+        # Create a Deferred with the above canceller, adding a reference to the
+        # deferred to the canceller to create the circular reference.
+        deferred: Deferred[Any] = Deferred(canceller)
+        canceller.deferred = deferred
+
+        deferred.callback(None)
+
+        del deferred
+        del canceller
+
+        # Once all local references have been dropped, the canceller should have
+        # been freed.
+        self.assertIsNone(weakCanceller())
+
+
+class _TestCircularCanceller:
+    """
+    c.f. L{DeferredTest.test_canceller_circular_reference}
+    """
+
+    deferred: Optional[Deferred[Any]] = None
+
+    def __call__(self, deferred: Deferred[Any]) -> None:
+        pass
```

Comment:
I think it's important to add the `pragma: no cover` marker if we don't expect this line to be called.

```suggestion
    def __call__(self, deferred: Deferred[Any]) -> None:  # pragma: no cover
        """
        This is not expected to be called as part of the current test suite.
        """
```

Reply:
Ah, didn't know about `# pragma: no cover`